cmd/auto-updater/main.go

```diff
@@ -75,6 +154,73 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 	return nil
 }
 
+func binaryVersion(location string) (ver version.SemVer, err error) {
+	out, err := exec.Command(location, "version").Output()
+	if err != nil {
+		return ver, err
+	}
+
+	scanner := bufio.NewScanner(bytes.NewReader(out))
+	for scanner.Scan() {
+		line := scanner.Text()
+		prefix := "Version: "
+		if strings.HasPrefix(line, prefix) {
+			return version.NewSemVer(line[len(prefix):])
+		}
+	}
+	return ver, nil
+}
+
+func suggestedVersion() (ver version.SemVer, url string, err error) {
+	resp, err := http.Get(versionURL)
+	if err != nil {
+		return ver, url, err
+	}
+	defer func() { err = errs.Combine(err, resp.Body.Close()) }()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return ver, url, err
+	}
+
+	var response Response
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		return ver, url, err
+	}
+
+	suggestedVersion := response.Processes.Storagenode.Suggested
+	ver, err = version.NewSemVer(suggestedVersion.Version)
+	if err != nil {
+		return ver, url, err
+	}
+	return ver, suggestedVersion.URL, nil
+}
+
+func downloadArchive(ctx context.Context, file io.Writer, url string) (err error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+
+	defer func() { err = errs.Combine(err, resp.Body.Close()) }()
+
+	if resp.StatusCode != http.StatusOK {
+		return errs.New("bad status: %s", resp.Status)
+	}
+
+	_, err = sync2.Copy(ctx, file, resp.Body)
+	return err
+}
+
+func fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
```

Comment:
It's safer to compare with `ModeType` to ensure that's a regular file and not any other like a symlink, etc.

Reply:
changed